### ISSUE TITLE
Minimisation of a mixed-type loss function

### ISSUE THREAD

**Issue by nhcho91** [SciML/DiffEqFlux.jl#656]
Hi. I am opening a new issue following the response to my question in [here](https://discourse.julialang.org/t/minimisation-of-a-bolza-form-loss-function-in-diffeqflux-framework/71920). For your convenience, I copy-pasted my original question below:

# Brief Summary
* Problem at hand:
minimise $J \left( p  \right) = \phi \left( x \left(t_{f} \right) \right) + \int_{t_{0}}^{t_{f}} L \left( \tau, x\left(\tau\right), p \right) d\tau + R\left( p \right)$
subject to $\dot{x}\left(t\right) = f\left( t, x\left(t\right), p\right)$
with $x\left(t_{0}\right) = x_{0}$, $t_{0}$, $t_{f}$  fixed,
where $p$ is the parameter for a neural network.

* How can I compute the correct gradient for the above Bolza form loss function (= continuous functional (Lagrange term) + terminal cost (Mayer term) + neural network parameter-dependent function ) through continuous adjoint sensitivity analysis using DiffEqSensitivity or DiffEqFlux?

* Is there any recommendation about the way of constructing the Bolza form loss function, e.g., including $\dot{J}_{cont}\left( t \right) = L\left( t, x\left(t\right), p \right)$ in the forward pass?

# More Details
Hello. My name is Namhoon Cho.

I was trying to solve a nonlinear deterministic optimal control problem in continuous-time domain by optimising a neural feedback policy. The context of my application requires specifying the terminal cost to be a function of final state in the form $\phi \left( x \left(t_{f}\right) \right)$ besides the continuous functional.

I have been following the codebase developed for models combining ODE and NNs.
At the first glance, I tried to just specify the loss function in such mixed form and apply DiffEqFlux.sciml_train (or others) for training.

However, I soon found out that the function adjoint_sensitivities() sitting inside DiffEqSensitivity.jl may not handle the loss function of a mixed type (continuous functional + terminal cost + parameter-only function,...) natively, which was also pointed out in [samuela/ctpg](https://github.com/samuela/ctpg/blob/a1973dedb38702076afacca7f54aad675b58ef0b/adjoint_test.jl#L12-L13). I read the description in [Mathematics of Sensitivity Analysis · DifferentialEquations.jl (sciml.ai)](https://diffeq.sciml.ai/stable/extras/sensitivity_math/), and the continuous adjoint sensitivity analysis described in CVODES document, but it is still not clear from my side how the current SciML packages handle the loss function term given by an evaluation at the final time.

Then, I became afraid that the result I obtain might not be based on the mathematically correct gradient $\frac{dJ}{dp}$ if I naively apply the existing package without knowing the details. 

Could anyone please clarify the thing whether I can apply the tools with no worries about correct $\frac{dJ}{dp}$ calculation?

Thank you.

**Comment by JinraeKim** [comment]
I've tried to make an example for your situation by augmenting the state to have a "cost state".
One might look like this, which is highly based on [an example code of DiffEqFlux.jl](https://diffeqflux.sciml.ai/stable/examples/neural_ode_flux/).
One thing that I haven't understood is that I have to assign `dx` terms as `Matrix` in the ODE function as the following.

```julia
function trueODEfunc(dx,x,p,t; a)
    true_A = [-0.1 2.0; -2.0 -0.1]
    dx[1:2, :] .= ((x[1:2, :] .^ 3)'true_A)' + a
    dx[3, :] = L(x[1:2, :])
end
```
- The example code
```julia
using DiffEqFlux, DifferentialEquations, Plots
using ComponentArrays
using FlightSims  # for apply_inputs

u0 = Float32[2.; 0.]
u0_aug = vcat(u0, zeros(1))
datasize = 30
tspan = (0.0f0,1.5f0)

L(u) = u' * u  # running cost

function trueODEfunc(dx,x,p,t; a)
    true_A = [-0.1 2.0; -2.0 -0.1]
    dx[1:2, :] .= ((x[1:2, :] .^ 3)'true_A)' + a
    dx[3, :] = L(x[1:2, :])
end
t = range(tspan[1],tspan[2],length=datasize)
# prob = ODEProblem(trueODEfunc,u0,tspan)
# prob = ODEProblem(trueODEfunc,u0_aug,tspan)
# ode_data = Array(solve(prob,Tsit5(),saveat=t))

# dudt2 = Chain(x -> x.^3,
#              Dense(2,50,tanh),
#              Dense(50,2))
# p,re = Flux.destructure(dudt2) # use this p as the initial condition!
# dudt(u,p,t) = re(p)(u) # need to restructure for backprop!
# prob = ODEProblem(dudt,u0_aug,tspan)
controller2 = Chain(u -> u.^3,
                   Dense(2,50,tanh),
                   Dense(50,2))
p,re = Flux.destructure(controller2)
controller(u, p, t) = re(p)(u)
prob = ODEProblem(apply_inputs(trueODEfunc; a=(x, p, t) -> controller(x[1:2], p, t)),u0_aug,tspan)

function predict_n_ode()
  Array(solve(prob,Tsit5(),u0=u0_aug,p=p,saveat=t))  # it seems that the output should be an Array
end

function loss_n_ode()
    pred = predict_n_ode()
    # loss = sum(abs2,ode_data .- pred)
    loss = pred[end][end]  # to make it scalar for gradient calculation
    loss
end

loss_n_ode() # n_ode.p stores the initial parameters of the neural ODE

cb = function (;doplot=false) # callback function to observe training
  pred = predict_n_ode()
  @show pred[end][end]
  # display(sum(abs2,ode_data .- pred))
  # # plot current prediction against data
  # pl = scatter(t,ode_data[1,:],label="data")
  # scatter!(pl,t,pred[1,:],label="prediction")
  # display(plot(pl))
  return false
end

# Display the ODE with the initial parameter values.
cb()

data = Iterators.repeated((), 1000)
# Flux.train!(loss_n_ode, Flux.params(u0,p), data, ADAM(0.05), cb = cb)
Flux.train!(loss_n_ode, Flux.params(p), data, ADAM(0.05), cb = cb)
```

- The result seems like:
```julia
(pred[end])[end] = 3.8348403741343478
(pred[end])[end] = 1.7142837453010644
(pred[end])[end] = 1.560227807289713
(pred[end])[end] = 1.2340005734359243
(pred[end])[end] = 1.008221684089493
(pred[end])[end] = 0.8959754967389809
(pred[end])[end] = 0.7389216617826677
(pred[end])[end] = 0.7074686470409919
(pred[end])[end] = 0.7379851618225052
(pred[end])[end] = 0.728009386563047
(pred[end])[end] = 0.7007186388029947
(pred[end])[end] = 0.6830207979108404
(pred[end])[end] = 0.6462518100394858
(pred[end])[end] = 0.5642265956053357
(pred[end])[end] = 0.44849382802697696
(pred[end])[end] = 0.3784813796477473
(pred[end])[end] = 0.41064424958072004
(pred[end])[end] = 0.46128482689398065
(pred[end])[end] = 0.46117894773364343
(pred[end])[end] = 0.4067221537342811
(pred[end])[end] = 0.3369709305899568
(pred[end])[end] = 0.30028202028455414
(pred[end])[end] = 0.3020057935440263
(pred[end])[end] = 0.31793688655564967
(pred[end])[end] = 0.32115284019145596
(pred[end])[end] = 0.322372187370472
(pred[end])[end] = 0.30715184511078536
(pred[end])[end] = 0.2699717255229833
(pred[end])[end] = 0.2382208463702111
(pred[end])[end] = 0.2523580097397865
(pred[end])[end] = 0.25722126648789445
(pred[end])[end] = 0.2520351715709875
(pred[end])[end] = 0.24668791390232053
(pred[end])[end] = 0.23354444746069575
(pred[end])[end] = 0.21866079851634415
(pred[end])[end] = 0.2207559316350696
(pred[end])[end] = 0.23448436126830935
(pred[end])[end] = 0.22305769601069111
(pred[end])[end] = 0.20230715849414735
(pred[end])[end] = 0.2055077955513695
(pred[end])[end] = 0.2129962298511977
(pred[end])[end] = 0.20875312911111946
(pred[end])[end] = 0.19820754949685374
(pred[end])[end] = 0.19233325469787288
(pred[end])[end] = 0.19461543675237178
(pred[end])[end] = 0.19738144791997494
(pred[end])[end] = 0.19208510038673632
(pred[end])[end] = 0.18248552690207379
(pred[end])[end] = 0.18401131155443218
(pred[end])[end] = 0.18661391306740135
(pred[end])[end] = 0.18076870106121198
(pred[end])[end] = 0.17752522654917652
(pred[end])[end] = 0.17569148071307186
(pred[end])[end] = 0.17574441820310588
(pred[end])[end] = 0.17490669698573566
(pred[end])[end] = 0.16924764155463554
(pred[end])[end] = 0.1688315381450853
(pred[end])[end] = 0.1698024712051132
(pred[end])[end] = 0.1665116354104631
(pred[end])[end] = 0.16408541427498535
(pred[end])[end] = 0.16312393663949068
(pred[end])[end] = 0.16323347691703843
(pred[end])[end] = 0.1609787605481437
(pred[end])[end] = 0.15838676766365847
(pred[end])[end] = 0.15895233606233486
(pred[end])[end] = 0.1574084027899759
(pred[end])[end] = 0.15573218145435688
(pred[end])[end] = 0.15437596068186274
(pred[end])[end] = 0.1540571413876138
(pred[end])[end] = 0.15281252078544452
(pred[end])[end] = 0.15092555251882575
(pred[end])[end] = 0.15067770958103407
(pred[end])[end] = 0.14961417693759968
(pred[end])[end] = 0.1483624308642799
(pred[end])[end] = 0.14715762292019377
(pred[end])[end] = 0.14684312842583053
(pred[end])[end] = 0.1454766616178978
(pred[end])[end] = 0.1444972327511784
(pred[end])[end] = 0.14374598442359937
(pred[end])[end] = 0.14300461010815943
(pred[end])[end] = 0.14175137262540619
(pred[end])[end] = 0.1411161918652191
(pred[end])[end] = 0.14031421675689557
(pred[end])[end] = 0.13937530631977063
(pred[end])[end] = 0.13847119892066403
(pred[end])[end] = 0.13791402109668538
(pred[end])[end] = 0.13691440816642483
(pred[end])[end] = 0.1361684353931602
(pred[end])[end] = 0.1354401972999217
(pred[end])[end] = 0.13470794645752285
(pred[end])[end] = 0.13383484219758215
(pred[end])[end] = 0.1332171175617558
(pred[end])[end] = 0.13247684404343488
(pred[end])[end] = 0.13167849308877855
(pred[end])[end] = 0.13106288791929785
(pred[end])[end] = 0.13033785361403938
(pred[end])[end] = 0.12963714898759046
(pred[end])[end] = 0.12897083148211583
(pred[end])[end] = 0.12832890759300167
(pred[end])[end] = 0.12764009892217523
(pred[end])[end] = 0.12699462115131566
(pred[end])[end] = 0.12639301367210737
(pred[end])[end] = 0.12571883770332998
(pred[end])[end] = 0.1251122878160372
(pred[end])[end] = 0.12452492159128688
(pred[end])[end] = 0.12388028179346677
(pred[end])[end] = 0.1233021829210286
(pred[end])[end] = 0.1227271492391227
(pred[end])[end] = 0.12211533608615564
(pred[end])[end] = 0.12155792228505383
(pred[end])[end] = 0.12099965938633733
(pred[end])[end] = 0.12041703108576005
(pred[end])[end] = 0.11987488142907086
(pred[end])[end] = 0.11933504043006721
(pred[end])[end] = 0.11878114455737036
(pred[end])[end] = 0.11825183605566127
(pred[end])[end] = 0.11772783041867596
(pred[end])[end] = 0.11720281850764376
(pred[end])[end] = 0.11668873810582979
(pred[end])[end] = 0.11617462282419197
(pred[end])[end] = 0.1156729841460958
(pred[end])[end] = 0.11518081536720565
(pred[end])[end] = 0.11468141022620164
(pred[end])[end] = 0.11419179702419456
(pred[end])[end] = 0.1137133215406556
(pred[end])[end] = 0.11323598412388237
(pred[end])[end] = 0.11276714612868234
(pred[end])[end] = 0.11230054549519548
(pred[end])[end] = 0.11183478168297736
(pred[end])[end] = 0.1113784116071918
(pred[end])[end] = 0.11092535747411036
(pred[end])[end] = 0.11047735075143496
(pred[end])[end] = 0.11003728032329253
(pred[end])[end] = 0.10959961729242733
(pred[end])[end] = 0.10916860492680903
(pred[end])[end] = 0.10874517668869008
(pred[end])[end] = 0.10832951025813259
(pred[end])[end] = 0.10793465109447321
(pred[end])[end] = 0.1075789162470039
(pred[end])[end] = 0.10734045587332915
(pred[end])[end] = 0.10734138294974276
(pred[end])[end] = 0.1081044598718331
(pred[end])[end] = 0.10869370539927192
(pred[end])[end] = 0.10983134043547513
(pred[end])[end] = 0.10670232333019064
(pred[end])[end] = 0.10530951035940678
(pred[end])[end] = 0.10454916755957076
(pred[end])[end] = 0.10414299767649092
(pred[end])[end] = 0.10393447285209229
(pred[end])[end] = 0.10434621514103742
(pred[end])[end] = 0.10565517820511955
(pred[end])[end] = 0.10872205095840468
(pred[end])[end] = 0.10381517625813491
(pred[end])[end] = 0.1020719045540015
(pred[end])[end] = 0.10141343091091233
(pred[end])[end] = 0.10106655710748505
(pred[end])[end] = 0.10093844989635473
(pred[end])[end] = 0.10158523863662373
(pred[end])[end] = 0.1049134455732741
(pred[end])[end] = 0.10431510699817584
(pred[end])[end] = 0.10421999471576912
(pred[end])[end] = 0.10039058470530135
(pred[end])[end] = 0.09931108374654644
(pred[end])[end] = 0.09904094921368371
(pred[end])[end] = 0.10004257252907155
(pred[end])[end] = 0.10226882371342039
(pred[end])[end] = 0.10560096559009989
(pred[end])[end] = 0.09752299040443968
(pred[end])[end] = 0.09865310083694467
(pred[end])[end] = 0.10752893241711603
(pred[end])[end] = 0.09644261849532565
(pred[end])[end] = 0.10108866470934104
(pred[end])[end] = 0.11141439836742234
(pred[end])[end] = 0.10011467199081103
(pred[end])[end] = 0.12713619609503155
(pred[end])[end] = 0.1139834665360549
(pred[end])[end] = 0.12453153851913926
(pred[end])[end] = 0.13069233387586782
(pred[end])[end] = 0.10863049097205493
(pred[end])[end] = 0.13884159449450484
(pred[end])[end] = 0.13889627310927244
(pred[end])[end] = 0.1103721931117587
(pred[end])[end] = 0.14547478184195664
(pred[end])[end] = 0.15773654178071708
(pred[end])[end] = 0.16427164772180125
(pred[end])[end] = 0.12782160607425316
(pred[end])[end] = 0.11310021941924876
(pred[end])[end] = 0.13873455552506642
(pred[end])[end] = 0.10706392067425968
(pred[end])[end] = 0.1260010614834865
(pred[end])[end] = 0.1514593095954888
(pred[end])[end] = 0.14199896035799314
(pred[end])[end] = 0.09654179415763114
(pred[end])[end] = 0.14342212251176012
(pred[end])[end] = 0.16887617044697828
(pred[end])[end] = 0.15460799393273128
(pred[end])[end] = 0.11581881814591417
(pred[end])[end] = 0.12761004751859914
(pred[end])[end] = 0.15778115296315168
(pred[end])[end] = 0.1634954735821293
(pred[end])[end] = 0.14974665652209293
(pred[end])[end] = 0.11404982802887788
(pred[end])[end] = 0.12389330578706315
(pred[end])[end] = 0.1453851075803709
(pred[end])[end] = 0.13907509700465762
(pred[end])[end] = 0.09718238927874524
(pred[end])[end] = 0.13199080117367393
(pred[end])[end] = 0.1558424195113213
(pred[end])[end] = 0.16265267287473884
(pred[end])[end] = 0.1399316139448252
(pred[end])[end] = 0.10106128815510423
(pred[end])[end] = 0.1356844096781582
(pred[end])[end] = 0.164714407244033
(pred[end])[end] = 0.1662535031665026
(pred[end])[end] = 0.14653559607920683
(pred[end])[end] = 0.09587599006702482
(pred[end])[end] = 0.13949256777122415
(pred[end])[end] = 0.17105452435224291
(pred[end])[end] = 0.1851348182966313
(pred[end])[end] = 0.18350968875343443
(pred[end])[end] = 0.165229914390469
(pred[end])[end] = 0.13358650575621767
(pred[end])[end] = 0.0844341038188681
(pred[end])[end] = 0.1322770390655041
(pred[end])[end] = 0.14422617000338114

```

**Comment by IlyaOrson** [comment]
I have used the Bolza form for the same purpose without inner modifications, here is [an example](https://github.com/IlyaOrson/control_neuralode/blob/juliacon2021/src/van_der_pol.jl#L90) and there are more in the [repo](https://github.com/IlyaOrson/control_neuralode).

**Comment by IlyaOrson** [comment]
This reference mentions further background on this (page 181) https://www.researchgate.net/publication/37452197_Nonlinear_and_Dynamic_Optimization_From_Theory_to_Practice
